# 🎬 Movie Recommendation System

## Phase 1: Data Understanding

### Objective

The objective of this notebook is to understand the movie dataset before preprocessing. This includes analyzing the dataset structure, checking missing values, identifying duplicate records, understanding data types, and making initial observations that will guide feature engineering and model development.

In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_colwidth", None)
pd.set_option("display.width", None)

In [2]:
df = pd.read_csv("../data/raw/movies.csv")

print("✅ Dataset Loaded Successfully!")

✅ Dataset Loaded Successfully!


In [3]:
df.head().T

,0,1,2,3,4
index,0,1,2,3,4
budget,237000000,300000000,245000000,250000000,260000000
genres,Action Adventure Fantasy Science Fiction,Adventure Fantasy Action,Action Adventure Crime,Action Crime Drama Thriller,Action Adventure Science Fiction
homepage,http://www.avatarmovie.com/,http://disney.go.com/disneypictures/pirates/,http://www.sonypictures.com/movies/spectre/,http://www.thedarkknightrises.com/,http://movies.disney.com/john-carter
id,19995,285,206647,49026,49529
keywords,culture clash future space war space colony society,ocean drug abuse exotic island east india trading company love of one's life,spy based on novel secret agent sequel mi6,dc comics crime fighter terrorist secret identity burglar,based on novel mars medallion space travel princess
original_language,en,en,en,en,en
original_title,Avatar,Pirates of the Caribbean: At World's End,Spectre,The Dark Knight Rises,John Carter
overview,"In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.","Captain Barbossa, long believed to be dead, has come back to life and is headed to the edge of the Earth with Will Turner and Elizabeth Swann. But nothing is quite as it seems.","A cryptic message from Bond’s past sends him on a trail to uncover a sinister organization. While M battles political forces to keep the secret service alive, Bond peels back the layers of deceit to reveal the terrible truth behind SPECTRE.","Following the death of District Attorney Harvey Dent, Batman assumes responsibility for Dent's crimes to protect the late attorney's reputation and is subsequently hunted by the Gotham City Police Department. Eight years later, Batman encounters the mysterious Selina Kyle and the villainous Bane, a new terrorist leader who overwhelms Gotham's finest. The Dark Knight resurfaces to protect a city that has branded him an enemy.","John Carter is a war-weary, former military captain who's inexplicably transported to the mysterious and exotic planet of Barsoom (Mars) and reluctantly becomes embroiled in an epic conflict. It's a world on the brink of collapse, and Carter rediscovers his humanity when he realizes the survival of Barsoom and its people rests in his hands."
popularity,150.437577,139.082615,107.376788,112.31295,43.926995


In [4]:
df.shape

(4803, 24)

In [5]:
df.columns

Index(['index', 'budget', 'genres', 'homepage', 'id', 'keywords',
       'original_language', 'original_title', 'overview', 'popularity',
       'production_companies', 'production_countries', 'release_date',
       'revenue', 'runtime', 'spoken_languages', 'status', 'tagline', 'title',
       'vote_average', 'vote_count', 'cast', 'crew', 'director'],
      dtype='str')

In [6]:
df.iloc[:5, :8]

,index,budget,genres,homepage,id,keywords,original_language,original_title
0,0,237000000,Action Adventure Fantasy Science Fiction,http://www.avatarmovie.com/,19995,culture clash future space war space colony society,en,Avatar
1,1,300000000,Adventure Fantasy Action,http://disney.go.com/disneypictures/pirates/,285,ocean drug abuse exotic island east india trading company love of one's life,en,Pirates of the Caribbean: At World's End
2,2,245000000,Action Adventure Crime,http://www.sonypictures.com/movies/spectre/,206647,spy based on novel secret agent sequel mi6,en,Spectre
3,3,250000000,Action Crime Drama Thriller,http://www.thedarkknightrises.com/,49026,dc comics crime fighter terrorist secret identity burglar,en,The Dark Knight Rises
4,4,260000000,Action Adventure Science Fiction,http://movies.disney.com/john-carter,49529,based on novel mars medallion space travel princess,en,John Carter


In [7]:
print("Number of Rows :", df.shape[0])
print("Number of Columns :", df.shape[1])

Number of Rows : 4803
Number of Columns : 24


In [8]:
missing_values = df.isnull().sum()

missing_values = missing_values[missing_values > 0]

missing_values.sort_values(ascending=False)

homepage        3091
tagline          844
keywords         412
cast              43
director          30
genres            28
overview           3
runtime            2
release_date       1
dtype: int64

In [9]:
print("Duplicate Rows :", df.duplicated().sum())

Duplicate Rows : 0


In [10]:
df.dtypes

index                     int64
budget                    int64
genres                      str
homepage                    str
id                        int64
keywords                    str
original_language           str
original_title              str
overview                    str
popularity              float64
production_companies        str
production_countries        str
release_date                str
revenue                   int64
runtime                 float64
spoken_languages            str
status                      str
tagline                     str
title                       str
vote_average            float64
vote_count                int64
cast                        str
crew                        str
director                    str
dtype: object

In [11]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
index,4803.0,2.401000e+03,1.386651e+03,0.0,1200.50000,2.401000e+03,3.601500e+03,4.802000e+03
budget,4803.0,2.904504e+07,4.072239e+07,0.0,790000.00000,1.500000e+07,4.000000e+07,3.800000e+08
id,4803.0,5.716548e+04,8.869461e+04,5.0,9014.50000,1.462900e+04,5.861050e+04,4.594880e+05
popularity,4803.0,2.149230e+01,3.181665e+01,0.0,4.66807,1.292159e+01,2.831350e+01,8.755813e+02
revenue,4803.0,8.226064e+07,1.628571e+08,0.0,0.00000,1.917000e+07,9.291719e+07,2.787965e+09
runtime,4801.0,1.068759e+02,2.261193e+01,0.0,94.00000,1.030000e+02,1.180000e+02,3.380000e+02
vote_average,4803.0,6.092172e+00,1.194612e+00,0.0,5.60000,6.200000e+00,6.800000e+00,1.000000e+01
vote_count,4803.0,6.902180e+02,1.234586e+03,0.0,54.00000,2.350000e+02,7.370000e+02,1.375200e+04


In [12]:
for i, column in enumerate(df.columns, start=1):
    print(f"{i}. {column}")

1. index
2. budget
3. genres
4. homepage
5. id
6. keywords
7. original_language
8. original_title
9. overview
10. popularity
11. production_companies
12. production_countries
13. release_date
14. revenue
15. runtime
16. spoken_languages
17. status
18. tagline
19. title
20. vote_average
21. vote_count
22. cast
23. crew
24. director


# Initial Observations

- The dataset contains **4,803 movies** with **24 attributes**.
- It includes both **numerical** and **categorical/textual** features.
- Important features include:
  - Title
  - Genres
  - Keywords
  - Cast
  - Director
  - Overview
  - Vote Average
  - Popularity
- Some columns contain missing values and will require preprocessing.
- The dataset does not contain movie poster images. Posters will later be fetched using the TMDb API.
- This dataset is well suited for building a **Content-Based Movie Recommendation System**.